src/layout_engine.rs

```diff
@@ -0,0 +1,206 @@
+	Vertical,
+}
+
+impl Axis {
+	/// Returns the axis perpendicular to this one.
+	fn perpendicular(&self) -> Axis {
+		match self {
+			Axis::Horizontal => Axis::Vertical,
+			Axis::Vertical => Axis::Horizontal,
+		}
+	}
+}
+/// Built-in component attributes used during layout.
+#[derive(Debug)]
+struct LayoutAttributes {
+	pub width: Dimension,
+	pub height: Dimension,
+}
+
+impl LayoutAttributes {
+	/// Retrieves the size of this element along the specified axis.
+	fn size_along(&self, axis: Axis) -> Dimension {
+		match axis {
+			Axis::Horizontal => self.width,
+			Axis::Vertical => self.height,
+		}
+	}
+
+	// fn alignment_along(axis: Axis)
+}
+
+/// A dimension is a measure along an axis.
+#[derive(Copy, Clone, Debug, PartialEq)]
+enum Dimension {
+	/// Absolute value in pixels.
+	AbsolutePx(f32),
+	/// Percent of parent container size along the same axis.
+	Percent(f32),
+	/// Percent of free space remaining in parent container.
+	PercentRemainder(f32),
+	/// Minimum size required to fit the children.
+	Inner,
+	/// Size relative to the width of this component.
+	Width,
+	/// Size relativ to the height of this component.
+	Height,
+}
+
+impl Dimension {
+	/// Converts an `AttributeValue` to a `Dimension`.
+	fn from_attribute_value(value: &AttributeValue) -> Self {
+		if let AttributeValue::TypeValue(values) = value {
+			assert_eq!(values.len(), 1, "Unexpected number of values");
+
+			if let TypeValueOrArgument::TypeValue(ref value) = values[0] {
+				let dimension = match *value {
+					TypeValue::AbsolutePx(pixels) => Dimension::AbsolutePx(pixels),
+					TypeValue::Percent(percent) => Dimension::Percent(percent),
+					TypeValue::PercentRemainder(percent) => Dimension::PercentRemainder(percent),
+					TypeValue::Inner => Dimension::Inner,
+					TypeValue::Width => Dimension::Width,
+					TypeValue::Height => Dimension::Height,
+					_ => panic!("unsupported type in layout attribute"),
+				};
+				dimension
+			}
+			else {
+				todo!("Arguments are not supported")
+			}
+		}
+		else {
+			todo!("variable parameters are not supported")
+		}
+	}
+}
+
+impl LayoutAbstractNode {
+	// TODO: when performing layout, we call this function multiple times.
+	// We should cache this data when a node is created.
+	fn layout_attributes(&self) -> LayoutAttributes {
+		match self {
+			Self::Tag(tag) => {
+				let width = tag.width().map(Dimension::from_attribute_value).unwrap_or(Dimension::Inner);
+				let height = tag.height().map(Dimension::from_attribute_value).unwrap_or(Dimension::Inner);
+
+				LayoutAttributes { width, height }
+			},
+			Self::Text(content) => {
+				// For now, we assume the text only takes up one row,
+				// and we have a monospaced font where each character is 20px wide.
+				let width = (content.len() * 20) as f32;
+				let height = 20.0;
+
+				let width = Dimension::AbsolutePx(width);
+				let height = Dimension::AbsolutePx(height);
+
+				LayoutAttributes { width, height }
+			},
```

Comment:
We won't even need to bother with this branch, text nodes aren't handled by the layout engine. They exist so you can pass text to a particular node's `content` attribute. The `<text>` primitive node will be in charge of rendering text and since it is `inner` by default, it will usually report its width when the text renderer reports that. The text (if you bind it, like `content="SOME_VAR"`) can also be used in the JavaScript(?) scripts associated with each layout component. For example, the `<dropdown>` element will need a list of selectable dropdown items so the JS will be `SOME_VAR.split("\n").map(s => s.trim())`.

Reply:
I've let it return width/height 100% for now so it doesn't crash my WIP cross-axis calculation code.